chapters/high_level_shader_language_comparison.adoc

```diff
@@ -0,0 +1,972 @@
+
+// @todo: maybe also add SPIR-V
+
+[TIP]
+====
+For more details on using HLSL with Vulkan, visit xref:{chapters}hlsl.adoc[this chapter].
+====
+
+[NOTE]
+====
+The following listings are by no means complete, and mappings for newer extensions may be missing. Also note that concepts do not always map 1:1 between GLSL and HLSL. E.g. there are no semantic in GLSL while some newer GLSL functionality may not (yet) be available in HLSL.
+====
+
+== Extensions
+
+In GLSL extensions need to be explicitly enabled using the `#extension` directive. This is **not** necessary in HLSL. The compiler will implicitly select suitable SPIR-V extensions based on the shader. If required one can use `-fspv-extension` arguments to explicitly select extensions.
+
+== Data types
+
+[NOTE]
+====
+Types work similar in GLSL and HLSL. But where GLSL e.g. has explicit vector or matrix types, HLSL uses basic types. On the other hand HLSL offers advanced type features like C++ templates. This paragraph contains a basic summary with some examples to show type differences between the two languages.
+====
+
+[options="header"]
+|====
+| *GLSL* | *HLSL* | *Example*
+| vec__n__ | float__n__ | vec4 -> float4
+| ivec__n__ | int__n__ | ivec3 --> int3
+| mat__nxm__ or shorthand mat__n__ | float__nxm__ | mat4 -> float4x4
+|====
+
+* link:https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-data-types[HLSL data types (Microsoft)]
+* link:https://www.khronos.org/opengl/wiki/Data_Type_(GLSL)[GLSL data types (OpenGL wiki)]
+
+The syntax for casting types also differs:
+
+GLSL:
+[source,glsl]
+----
+mat4x3 mat = mat4x3(ubo.view);
+----
+
+HLSL:
+[source,hlsl]
+----
+float4x3 mat = (float4x3)(ubo.view);
+----
+
+== Built-ins vs. Semantics
+
+[NOTE]
+====
+While GLSL makes heavy use of input and output variables built into the languages called "built-ins", there is no such concept in HLSL. HLSL instead uses link:https://learn.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-semantics[semantics], strings that are attached to inputs or inputs that contain information about the intended use of that variable. They are prefixed with `SV_`. For HLSL input values are explicit arguments for the main entry point and the shader needs to explicitly return an output.
+====
+
+=== Examples
+
+Writing positions from the vertex shader:
+
+GLSL:
+[source,glsl]
+----
+layout (location = 0) in vec4 inPos;
+
+void main() {
+    // The vertex output position is written to the gl_Position built-in
+    gl_Position = ubo.projectionMatrix * ubo.viewMatrix * ubo.modelMatrix * inPos.xyz;
+}
+----
+
+HLSL
+[source,hlsl]
+----
+struct VSOutput
+{
+    // The SV_POSITION semantic declares the Pos member as the vertex output position 
+	float4 Pos : SV_POSITION;
+};
+
+VSOutput main(VSInput input)
+{
+	VSOutput output = (VSOutput)0;
+	output.Pos = mul(ubo.projectionMatrix, mul(ubo.viewMatrix, mul(ubo.modelMatrix, input.Pos)));
```

Comment:
Normally, matrix multiples in HLSL tend to be written in opposite order due to the weirdness of row vs col major. Default buffer layout is col-major, so there's an implicit transpose.

Reply:
It seems that this is implicitly done by DXC when using the SPIR-V target.